Code/GraphMol/Descriptors/WHIM.cpp

```diff
@@ -0,0 +1,625 @@
+
+
+using namespace Eigen;
+
+#define EIGEN_TOLERANCE 1.0e-2
+
+
+namespace RDKit {
+namespace Descriptors{
+
+namespace {
+
+  MolData3Ddescriptors moldata3D;
+
+
+double roundn(double in,int factor) {
+
+  return round(in*pow(10,factor))/pow(10,factor);
+}
+
+
+double* retreiveMat(Eigen::MatrixXd matrix) {
+   double* arrayd = matrix.data();
+   return arrayd;
+}
+
+double* retreiveVect(Eigen::VectorXd matrix) {
+   double* arrayd = matrix.data();
+   return arrayd;
+
+}
+
+
+MatrixXd GetCenterMatrix(MatrixXd Mat){
+
+    VectorXd v = Mat.colwise().mean();
+
+    MatrixXd X=Mat.rowwise() - v.transpose();
+
+    return X;
+}
+
+MatrixXd GetCovMatrix(MatrixXd X, MatrixXd Weigth, double weigth){
+
+    return X.transpose() * Weigth * X / weigth;
+}
+
+
+JacobiSVD<MatrixXd> getSVD(MatrixXd Mat) {
+
+    JacobiSVD<MatrixXd> svd(Mat,  ComputeThinU | ComputeThinV);
+
+    return svd;
+}
+
+
+
+double getKu(double* w, int numAtoms) {
+
+  double res=0.0;
+  double sums=0.0;
+  for (int i=0;i<numAtoms;i++) 
+  {
+      sums+=w[i];
+  }
+
+  for (int i=0;i<numAtoms;i++) 
+  {
+    res+=std::abs(i/sums-1.0/3.0);
+  }
+
+  double Ku = 3.0/4.0*res;
+  return round(1000*Ku)/1000;
+}
+
+double* getWhimDesc(JacobiSVD<MatrixXd> svd, MatrixXd Xmean, int numAtoms, double th,bool printscore) {
+
+    double *w = new double[18];
+    // prepare data for Whim parameter computation
+    
+    double *SingVal = retreiveVect(svd.singularValues());
+    MatrixXd Scores= Xmean*svd.matrixV(); //  V is similar
+
+    // compute parameters
+    w[0]=SingVal[0];
+    w[1]=SingVal[1];
+    w[2]=SingVal[2];
+    w[3]=SingVal[0]+SingVal[1]+SingVal[2];  // T
+    w[4]=SingVal[0]*SingVal[1]+SingVal[0]*SingVal[2]+SingVal[1]*SingVal[2]; // A
+    w[5]=w[3]+w[4]+SingVal[0]*SingVal[1]*SingVal[2]; // V
+    w[6]=SingVal[0]/(SingVal[0]+SingVal[1]+SingVal[2]); // P1
+    w[7]=SingVal[1]/(SingVal[0]+SingVal[1]+SingVal[2]); // p2
+    w[8]=SingVal[2]/(SingVal[0]+SingVal[1]+SingVal[2]); // P3
+
+    double res=0.0;
+    for (int i=0;i<3;i++) 
+    {
+      res+=std::abs(w[i]/w[3]-1.0/3.0);
+    }
+
+    w[9]=3.0/4.0*res; // K
+
+    // center original matrix
+
+    VectorXd v1= Scores.col(0);
+    VectorXd v2= Scores.col(1);
+    VectorXd v3= Scores.col(2);
+
+    if (printscore and th ==0.03) {
+          std::cout << v1.transpose() << "\n";
+          std::cout << v2.transpose() << "\n";
+          std::cout << v3.transpose() << "\n";
+    }
+
+    w[10] = numAtoms*pow(w[0],2)/ v1.array().pow(4).sum(); // E1
+    w[11] = numAtoms*pow(w[1],2)/ v2.array().pow(4).sum(); // E2
+    w[12] = numAtoms*pow(w[2],2)/ v3.array().pow(4).sum(); // E3
+    w[13] = w[10]+w[11]+w[12]; //D
+
+
+    double gamma[3]; // Gamma values
+    double nAT = (double) numAtoms;
+    // check if two atoms are symetric versus the new axis ie newx,newy,newz a
+    for (int i=0;i<3; i++) {
+      for (int j =0;j<numAtoms;j++) {
+        Scores(j,i)=roundn(Scores(j,i),2); // round the matrix! same as eigen tolerance !
```

Comment:
we can have lot of rounding issues (deviation of result vs Dragon). We can also consider not to round and saying that we are not similar to Dragon... I cannot decide what is the best solution.
But based on those code I will prefer the second option cause, We cannot be exactly identical to Dragon, We can try to reproduce paper's but for some cases It's clear that not all is write in the paper for some of those descriptors!